### DISCUSSION TITLE
Why does AIR limit GPU upload size?

### DISCUSSION THREAD

**Discussion by eyal-katz** [airsdk/Adobe-Runtime-Support#2143] [General]
Today games need more than 512MB of GPU and that’s really putting a threat on our application.

Do you know if there’s any rational cause to limit the GPU upload size?

How can this be changed?

**Comment by eyal-katz** [comment]
From [Context3D API Reference](https://help.adobe.com/en_US/FlashPlatform/reference/actionscript/3/flash/display3D/Context3D.html)
> 512 MB is the absolute limit for textures, including the texture memory required for mipmaps. However, for Cube Textures, the memory limit is 256 MB.

**Reply by marchbold** [reply]
Good question, @ajwfrost any insights?

**Reply by ajwfrost** [reply]
Just spoke with the Adobe team on this, to confirm -> yes per @matheusdiasdesouzads 's suggestion, it was something to try to get parity between the different devices and capabilities. So something that should be being updated as devices get more capable -> or something where we need to ensure that the developers are able to handle errors gracefully. So e.g. what might the behaviour be if you're trying to load your game assets but the GPU runs out of texture storage ...  if we just through some AS3 exception for this, then it's down to the game developer to handle the condition and .. do whatever needs doing to recover gracefully!

@eyal-katz  we can look at how best to do this; whether it's a new profile or whether it's just some setting that can be added somehow (like an app descriptor flag, or a new property on the Context3D instance, etc...) - the simpler the better, probably, but without impact existing applications!

thanks

**Reply by eyal-katz** [reply]
@ajwfrost thanks for the elaborated response.
I think that for a quick action - let us raise a flag to support 1GB instead of 500MB. That will release the immediate stress we’re facing, and perhaps won’t impact others too much, what do you think?

**Reply by marchbold** [reply]
I'd probably suggest a new profile for this one? I believe this is how adobe approached it in the past.

**Reply by ajwfrost** [reply]
Good suggestion .. would it be fair to say people are mostly using `ENHANCED` at this point, i.e. we could create `ENHANCED_EXTENDED`? Or do we need `BASELINE_EVEN_MORE_EXTENDED`?! .. `BASELINE_UNLIMITED` is something I could see i.e. there's a level of API support (`BASELINE`/`STANDARD`/`ENHANCED`) and there's a level of how much memory/resources to allow (`CONSTRAINED`/normal/`EXTENDED`)...
Or we split these out into a profile/API feature set, and a separate flag/level for the resource limits?

**Reply by marchbold** [reply]
Starling uses them in this order: 

```
"enhanced",
"standardExtended", "standard", "standardConstrained",
"baselineExtended", "baseline", "baselineConstrained"
```

Most of our apps will be using `ENHANCED`, so I'd say `ENHANCED_EXTENDED` is a good approach.

**Reply by hydroper** [reply]
`enhanceExtended` for at least 2GB? 1GB is still little, I think.

**Reply by eyal-katz** [reply]
So maybe just add a flag saying “don’t limit my gpu access I know what I’m doing”? The flag is profile agnostic, just tells AIR not to check limits… why isn’t this sufficient?

**Reply by ajwfrost** [reply]
@eyal-katz we were just looking at this and found we'd already made a change here (!) ... if you publish your app for 33.1 (or above) then we should already be ignoring the resource limits, in terms of the maximum total count of and memory used by things like programs, vertex/index buffers, and textures.

Leaving aside the sense of just dropping the limits like we appear to have done, can I check whether you're seeing these restrictions still?

Or is it another restriction you're looking at, e.g. the "maximum size" for an individual texture?

**Reply by eyal-katz** [reply]
@ajwfrost I will check (ASAP) if the restrictions were dropped on 33.1 and let you know.

**Reply by eyal-katz** [reply]
@ajwfrost 
It doesn't look so good
I got Error #3691 after uploading 770MB to GPU
Tested on iPhone 13 Pro 
AIR Version: 33,1,1,926 
I ran this line of code every 1 second:
`Texture.fromBitmapData(new BitmapData(512, 512, false, GREEN)); // 1MB in GPU`

**Reply by ajwfrost** [reply]
Strange for this to happen after 770MB?! The error 3691 should only be sent when we've detected a soft-limit i.e. the restrictions that were added, but 770MB isn't one of the levels that should trigger things.. if the device had run out of memory then a different error should have been thrown.

FYI we've added a more explicit mechanism for this in version 50.0 but if the above scenario throws the error, I can't see how the new mechanism would avoid it. Will have to check on this a little more....

thanks

**Reply by eyal-katz** [reply]
@ajwfrost - The plot thickens:

On the same iPhone (13 Pro) I got Error #3691 after uploading ATF textures, when the GPU reported around 550MB
to the GPU (Not BitmapData, which's limit was 770MB).

This was also the case on two different Android devices (Galaxy 21 and One Plus)

**But** - on a [Realme 6 Pro](https://www.gsmarena.com/realme_6_pro-10114.php) there were no limits when uploading ATF or BitmapData.

All devices worked with work with AIR 33,1,1,926

**Comment by hydroper** [comment]
Maybe most users resources were more limited when 3D display features were being developed in Flash Player 11? Nowadays 2GB+ graphics card are normal (my laptop, which is simple, has a 4GB graphics card).

**Comment by eyal-katz** [comment]
How can we ask to raise/cancel the limit ?

**Comment by bwhiting** [comment]
Steam publishes monthly stats and you can check out the vram entries here.
https://store.steampowered.com/hwsurvey/Steam-Hardware-Software-Survey-Welcome-to-Steam
Obviously this is a gaming audience but it's one source I could find easily at least. Not geographical or demographical broken down or anything but still could be useful.

**Reply by hydroper** [reply]
Wow... didn't know 8GB was most common for VRAM.

**Comment by blackjyn** [comment]
UP!

512MB limit nowadays is nonsense

**Comment by RossD20Studios** [comment]
I know it's been awhile since this ticket was last commented on, but a fellow developer drew my attention here, and so I wanted to chime in that I, too, would love to see the resource limits (if any imposed) increased. Several of my players are pushing 6K+ monitors and being limited to 4K textures is capping a potentially breathtaking visual experience that could be solved with 8K textures and higher memory limits.

**Comment by Sephius** [comment]
In this advanced era of 2023, it is indeed quite perplexing to encounter a 512MB limit for texture usage, particularly when we reflect on modern technological capabilities. While I can understand that there may have been reasons to enforce this limit in the past, it seems anachronistic in today's context.

When it comes to platform considerations, I must respectfully assert that this constraint never seemed justifiable. In my view, it would be more reasonable to leave memory-related decisions to the discretion of developers, who are typically more attuned to the specific needs and constraints of the target device they are developing an app or game for.

This lingering limitation does not only serve as an inconvenience; it continues to be a troublesome legacy that affects our present-day development.

Furthermore, I feel compelled to express some concerns regarding Harman's SDK. It appears to present additional challenges, as it seems to enforce a similar memory RAM limit of 512MB. This restriction, unfortunately, leads to app crashes even when loading bytes for later upload to the GPU.

I highlight these issues not to disparage but to emphasize areas where thoughtful improvement could greatly benefit all involved. I hope that attention to these matters will lead to more robust solutions that are in line with our shared goals for advancement and innovation.

**Comment by Sephius** [comment]
I just found a way to use an unlimited amount of GPU memory:
- In a Context3D object there is a variable named: [ignoreResourceLimits (https://airsdk.dev/reference/actionscript/3.0/flash/display3D/Context3D.html#ignoreResourceLimits)

If is set to TRUE app will no longer crash after 512mb is reached. I don't know what is the reason for this to be true by default, but it really removes the limitation.

I was able to upload more than 3GBs of textures after this.

**Reply by blackjyn** [reply]
Thanks 👍
